### PR TITLE
feat: create lambda push/sendNotificationToDevice

### PR DESCRIPTION
**Pull request by alexruzenhack**
### Acceptance Criteria
- create lambda push/sendNotificationToDevice
- create utility to invoke sendNotificationToDevice

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/api/pushSendNotificationToDevice.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import { unregisterPushDevice } from '@src/db';
+
+const mysql = getDbConnection();
+
+class PushSendNotificationToDeviceInputValidator {
+  static readonly bodySchema = Joi.object({
+    deviceId: Joi.string().max(256).required(),
+    title: Joi.string().required(),
+    description: Joi.string().required(),
+    metadata: Joi.object().optional(),
```
Comment (andreabadesso) [inline src/api/pushSendNotificationToDevice.ts]: Should we describe better this object? What is expected?

File: src/db/index.ts [hunk 2]
```diff
@@ -2762,14 +2763,23 @@ export const updatePushDevice = async (
...
-    `
-    DELETE FROM \`push_devices\`
-    WHERE device_id = ? AND wallet_id = ?`,
-    [deviceId, walletId],
-  );
+  if (walletId) {
+    await mysql.query(
+      `
+      DELETE FROM \`push_devices\`
+      WHERE device_id = ? AND wallet_id = ?`,
```
Comment (andreabadesso) [inline src/db/index.ts]:
```suggestion
      DELETE
        FROM \`push_devices\`
       WHERE \`device_id\` = ?
         AND \`wallet_id\` = ?`,
```

File: src/db/index.ts [hunk 3]
```diff
@@ -2840,3 +2850,31 @@ export const existsWallet = async (
...
+ *
+ * @param mysql - Database connection
+ * @param deviceId - The device to verify existence
+ */
+export const getPushDevice = async (
+  mysql: ServerlessMysql,
+  deviceId: string,
+) : Promise<PushDevice|null> => {
+  const [pushDevice] = await mysql.query(
+    'SELECT * FROM `push_devices` WHERE device_id = ?',
```
Comment (andreabadesso) [inline src/db/index.ts]:
```suggestion
`SELECT *
   FROM \`push_devices\`
  WHERE \`device_id\` = ?`,
```

File: src/utils/pushnotification.utils.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { Lambda } from 'aws-sdk';
+import { SendNotificationToDevice } from '@src/types';
+
+const SEND_NOTIFICATION_LAMBDA_ENDPOINT = process.env.SEND_NOTIFICATION_LAMBDA_ENDPOINT;
```
Comment (andreabadesso) [inline src/utils/pushnotification.utils.ts:15]: Maybe we should throw an error here or log with `ALERT` so we get notified that this env is not set

File: src/utils/pushnotification.utils.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+const SEND_NOTIFICATION_FUNCTION_NAME = `hathor-wallet-service-${process.env.STAGE}-sendNotificationToDevice`;
+
+export enum PushNotificationError {
+  INVALID_DEVICE_ID = 'invalid-device-id',
+  UNKNOWN = 'unknown',
+}
+
+export class PushNotificationUtils {
+  static sendToFcm(_notification: SendNotificationToDevice): Promise<{ success: boolean, errorMessage?: string }> {
+    return Promise.race([]);
```
Comment (andreabadesso) [inline src/utils/pushnotification.utils.ts:26]: What is this doing?

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/api/pushSendNotificationToDevice.ts (hunk 1 above)
Comment (andreabadesso) [reply src/api/pushSendNotificationToDevice.ts]: Btw, is this even being used?

Comment (alexruzenhack) [reply src/api/pushSendNotificationToDevice.ts]: @tuliomir Can you clarify for what this `metadata` will be useful?

File: src/utils/pushnotification.utils.ts (hunk 5 above)
Comment (alexruzenhack) [reply src/utils/pushnotification.utils.ts:26]: Oh... this was just a placeholder for the next PR.

Comment (alexruzenhack) [reply src/utils/pushnotification.utils.ts:26]: You can see the implementation here:
https://github.com/HathorNetwork/hathor-wallet-service/blob/6d0515da45e2ce022de2f066c4cc8d2b43212415/src/utils/pushnotification.utils.ts#L22-L43

**Review by andreabadesso — APPROVED** [review]

File: src/api/pushSendNotificationToDevice.ts (hunk 1 above)
Comment (tuliomir) [reply src/api/pushSendNotificationToDevice.ts]: We can send JSON metadata with the push. In a previous iteration of the design, there was the possibility of sending an entire transaction here.

Now we only use it for sending the `txId` of the relevant transaction, for example:
```json
"metadata": { 
  "txId": "00e2597222154cf99bfef171e27374e7f35aa7448afae10c15e9f049b95a3e67"
}
```

File: src/utils/pushnotification.utils.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (tuliomir) [inline src/utils/pushnotification.utils.ts:26]: I think this deserves a `TODO` or some other kind of code comment to clarify it's yet to be implemented

File: tests/pushSendNotificationToDevice.test.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    enableShowAmounts: false,
+  });
+  await register(event, null, null) as APIGatewayProxyResult;
+
+  const validPayload = {
+    deviceId: 'device1',
+    title: 'You have a new transaction',
+    description: '5HTR was sent to my-wallet',
+    metadata: {
+      firstMetadata: 'firstMetadata',
```
Comment (tuliomir) [inline tests/pushSendNotificationToDevice.test.ts]: This test could have the sample metadata content expected on the [front-end design](https://github.com/HathorNetwork/internal-issues/issues/132), for example:
```json
"metadata": { 
  "txId": "00e2597222154cf99bfef171e27374e7f35aa7448afae10c15e9f049b95a3e67"
}
```

File: tests/pushSendNotificationToDevice.test.ts [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+  makeGatewayEventWithAuthorizer,
+  cleanDatabase,
+  checkPushDevicesTable,
+} from '@tests/utils';
+import { APIGatewayProxyResult, Context } from 'aws-lambda';
+
+const mysql = getDbConnection();
+
+const spyOnSendToFcm = jest.spyOn(PushNotificationUtils, 'sendToFcm');
+const spyOnLoggerError = jest.spyOn(logger, 'error');
```
Comment (tuliomir) [inline tests/pushSendNotificationToDevice.test.ts:24]: Good point! I think we didn't have tests for the logger yet.

**Review by tuliomir — APPROVED** [review]

**Pull request merged**